### DISCUSSION TITLE
add IWYU pragma to internal headers

### DISCUSSION THREAD

**Discussion by maldag** [stephenberry/glaze#1163] [General]
Using LLVM tools for compilation (clang++) and for static analysis (clang-tidy) the check `misc-include-cleaner` doesn't know about private or public headers of glaze library.

According to <https://clangd.llvm.org/design/include-cleaner#umbrella-headers> private headers should add the following pragma at the beginning of private header files:

`// IWYU pragma: private, include "glaze/glaze.hpp"`

This seems to come from here: <https://github.com/include-what-you-use/include-what-you-use>

I would be happy if this could be included in private headers that users should not use directly

**Comment by stephenberry** [comment]
Thanks for bringing this up. I've worked on projects in the past that have tried to use IWYU, but we ran into too many mistakes and it made the included headers uglier without understanding that nested includes are fine when aggregated for a particular translation unit. Hopefully it's gotten better since then.

I'm a little wary about trying to support `misc-include-cleaner` because of how Glaze works with template specializations, where relationships are not direct function calls.

Is there a way to just tell IWYU to ignore Glaze files?

Anyway, I am open to a pull request that adds this if you would like to submit one.

Also, I'm hoping to add C++20 module support to Glaze in the near future, which would make IWYU and this check obsolete.

I'm going to move this to discussions because it isn't high priority, but I am interested in this and let me know if there's a simple solution.

Thanks!
